In [ ]:
# Section 1: Numerical Data Types & Arithmetic Operators
# The Road Trip Planner

In [35]:
# a. How many liters of fuel will you need for the entire trip?
distance = 475  # kilometers
fuel_efficiency = 11.5  # kilometers per liter
fuel_needed = distance / fuel_efficiency
print(f"Fuel needed: {fuel_needed} liters")

Fuel needed: 41.30434782608695 liters


To calculate the fuel needed, I divided the total distance by the fuel efficiency. This gives us the total liters required to travel 475 kilometers at an efficiency of 11.5 km/L.

In [36]:
# b. How many full tanks of fuel does this represent?
tank_capacity = 50  # liters
full_tanks = fuel_needed // tank_capacity
print(f"Full tanks needed: {full_tanks}")

Full tanks needed: 0.0


I used the floor division operator (//) to find how many complete tanks would be needed. Since 41.3 liters is less than a single tank's capacity (50 liters), the result is 0 full tanks. This means we can complete the trip with just the initial full tank.

In [37]:
# c. What is the total estimated fuel cost for the trip?
fuel_price = 1.65  # dollars per liter
fuel_cost = fuel_needed * fuel_price
print(f"Total fuel cost: ${fuel_cost:.2f}")

Total fuel cost: $68.15


To calculate the total fuel cost, I multiplied the amount of fuel needed (41.3 liters) by the price per liter ($1.65). This gives the total cost of the fuel required for the trip.

In [38]:
# d. How many liters of fuel will be remaining in your tank?
starting_fuel = 50  # liters (full tank)
remaining_fuel = starting_fuel - fuel_needed
print(f"Remaining fuel: {remaining_fuel} liters")

Remaining fuel: 8.695652173913047 liters


Since we start with a full tank of 50 liters and use 41.3 liters for the trip, we'll have approximately 8.7 liters remaining. This makes sense because the fuel needed was less than a full tank.

In [ ]:
# Section 2: Comparison & Logical Operators
# Theme Park Entry Rules

In [39]:
# a. Write a comprehensive boolean expression for the discount qualification
age = 30
is_weekday = True
is_family_day = False
group_size = 3

qualifies_for_discount = (age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)
print(f"Qualifies for discount: {qualifies_for_discount}")

Qualifies for discount: False


The expression combines the three discount conditions using logical operators. With the current values (age 30, weekday, not family day, group size 3), none of the conditions are met, so the result is False.

In [40]:
# b. Evaluate the expression with new values
age = 10
is_weekday = True
is_family_day = False
group_size = 2

qualifies_for_discount = (age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4)
print(f"Qualifies for discount: {qualifies_for_discount}")

Qualifies for discount: True


# c. Explanation of evaluation Step-by-step
age = 10
is_weekday = True
is_family_day = False
group_size = 2
First condition: (age <= 12 and is_weekday)
•	age <= 12 is 10 <= 12, which is True
•	is_weekday is True
•	True and True evaluates to True
Second condition: (age >= 65)
•	age >= 65 is 10 >= 65, which is False
Third condition: (is_family_day and group_size >= 4)
•	is_family_day is False
•	group_size >= 4 is 2 >= 4, which is False
•	False and False evaluates to False
Combining all conditions: True or False or False
•	True or False evaluates to True
•	True or False evaluates to True
Since the first condition is True (child on a weekday), the entire expression evaluates to True, meaning the person qualifies for a discount.

In [ ]:
Section 3: String Data Types, Indexing & Slicing
Message Decoder

In [41]:
# a. Character at index 5
secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"
char_at_index_5 = secret_message[5]
print(f"Character at index 5: '{char_at_index_5}'")

Character at index 5: 'N'


In Python, string indexing starts at 0, so the 6th character (at index 5) is 'N'.

In [42]:
# b. Extracting "Is_FuN" using positive indexing
substring_is_fun = secret_message[7:14]
print(f"Substring 'Is_FuN': '{substring_is_fun}'")

Substring 'Is_FuN': 'Is_FuN_'


To extract "Is_FuN", I identified that it starts at index 7 and ends before index 14. In Python slicing notation, this is written as [7:14].

In [43]:
# c. Extracting "PoWeRfUl" using negative indexing
substring_powerful = secret_message[-9:-1]
print(f"Substring 'PoWeRfUl': '{substring_powerful}'")

Substring 'PoWeRfUl': 'PoWeRfUl'


Using negative indexing, I count from the end of the string. "PoWeRfUl" starts at the 9th character from the end (index -9) and ends before the last character (index -1).

In [44]:
# d. Extract every 4th character
every_fourth = secret_message[0::4]
print(f"Every 4th character: '{every_fourth}'")

Every 4th character: 'PosNDWU'


# e. Accessing secret_message[50].
If we try to access secret_message[50], we'll get an IndexError: string index out of range error.
This error occurs because the string doesn't have an index 50. The length of secret_message is 29 characters, so valid indices range from 0 to 28 (or -1 to -29 for negative indexing). Trying to access an index outside this range results in an IndexError, which is google colab or python way of telling us we're attempting to access a position that doesn't exist.


In [ ]:
# Section 4: Boolean Values & Expressions
# Predict and Verify

In [23]:
# Given Values
x = 15
y = 25
z = 15
text = "adventure"

In [24]:
# a. (x == z) and (y > x)
result_a = (x == z) and (y > x)
print(f"(x == z) and (y > x) = {result_a}")

(x == z) and (y > x) = True


•	x == z is 15 == 15, which is True
•	y > x is 25 > 15, which is True
•	True and True evaluates to *True*

In [25]:
# b. (x != z) or (y < x) or (len(text) < 5)
result_b = (x != z) or (y < x) or (len(text) < 5)
print(f"(x != z) or (y < x) or (len(text) < 5) = {result_b}")


(x != z) or (y < x) or (len(text) < 5) = False


•	x != z is 15 != 15, which is False
•	y < x is 25 < 15, which is False
•	len(text) < 5 is 9 < 5, which is False
•	False or False or False evaluates to False

In [28]:
# c. not ((x > y) and (text == "Adventure")) # This is not valid Python syntax
result_c = not ((x > y) and (text == "Adventure"))
print(f"not ((x > y) and (text == \"Adventure\")) = {result_c}")

not ((x > y) and (text == "Adventure")) = True


•	x > y is 15 > 25, which is False
•	text == "Adventure" is "adventure" == "Adventure", which is False (case-sensitive)
•	False and False evaluates to False
•	not False evaluates to True

In [ ]:
# Section 5: String Properties & Basic Methods
# User Profile Setup

In [29]:
# a. Changing the first letter of a username
username = "MyUser123"
try:
    username[0] = 'm'
    print(username)
except TypeError as e:
    print(f"Error: {e}")

Error: 'str' object does not support item assignment


This code results in a TypeError because strings in Python/google colabare immutable. This means once a string is created, you cannot change its individual characters. To modify a string, you must create a new string with the desired changes.

In [31]:
# b. Transforming a quote
quote = " practice makes perfect "
# Step 1: Remove leading/trailing whitespace
quote_trimmed = quote.strip()
print(f"After strip(): '{quote_trimmed}'")

After strip(): 'practice makes perfect'


The strip() method removes any leading and trailing whitespace from the string.

In [32]:
# Step 2: Convert to uppercase
quote_upper = quote_trimmed.upper()
print(f"After upper(): '{quote_upper}'")

After upper(): 'PRACTICE MAKES PERFECT'


The upper() method converts all characters in the string to uppercase.

In [33]:
# Step 3: Replace "PERFECT" with "PROGRESS"
formatted_quote = quote_upper.replace("PERFECT", "PROGRESS")
print(f"Final formatted quote: '{formatted_quote}'")

Final formatted quote: 'PRACTICE MAKES PROGRESS'


 The replace() method substitutes all occurrences of the first argument with the second argument.

In [34]:
# c. Finding the lengths of the quotes
original_length = len(quote)
formatted_length = len(formatted_quote)

print(f"Length of original quote: {original_length} characters")
print(f"Length of formatted quote: {formatted_length} characters")

Length of original quote: 24 characters
Length of formatted quote: 23 characters


The len() function returns the number of characters in a string, including spaces. The original quote includes leading and trailing spaces, making it longer than the formatted quote which has the spaces removed (although it has a longer replacement word "PROGRESS" instead of "PERFECT").